### DISCUSSION TITLE
Formatting/Transforming Variables (Attempting to Inverse/Bitwise Not variables)

### DISCUSSION THREAD

**Discussion by Th3Alic3** [WerWolv/ImHex#688] [Get Help]
Trying to display at set of X and Y coords that are stored inverted/Bitwise NOT.

Tried this:

struct SpriteDimensionValues{
	u16 X [[~a]];
	u16 Y [[~a]];
};

Tried this:

struct SpriteDimensionValues{
	u16 X [[format("~a")]];
	u16 Y [[format("~a")]];
};

and tried this:

struct SpriteDimensionValues{
	u16 X [[transform("~a")]];
	u16 Y [[transform("~a")]];
};

and finally tried this:

struct SpriteDimensionValues{
	u16 ~X;
	u16 ~Y;
};

The error just keeps saying "~a" doesn't exist but it's listed in the expressions (https://imhex.werwolv.net/docs/core_language/expressions.html).

@WerWolv in response to adding a feature request not to long ago indicated both the [[format]] and [[transform]] should work to translate the values from the inverted data. However, I cannot find anything in documentation on either of these tools to apply within the pattern. Since both state they require a defined name and neither accept the math expression of Bitwise Not... I am at a loss how to actually impliment this within my pattern...

I wholeheartly believe this is user error and that the app can support the needs... just need someone to point out the actually usage for the Pattern to do its thing. :D

**Comment by Simran-B** [comment]
`format()` and `transform()` expect the name of a function to call and in that function, you can use the `~` operator (but note that this seems to use 64-bit internally and thus needs to be re-casted):

```elixir
struct SpriteDimensionValues {
  u16 X [[format_read("invert"), format_write("invert")]];
  u16 Y [[format_read("invert"), format_write("invert")]];
};

fn invert(u16 val) {
  return u16(~val);
};

SpriteDimensionValues v @ 0x00;
```

**Reply by paxcut** [reply]
This is a very old question but it should be answered correctly in case someone tries to get help reading this.I don't think `format_read` and `format_write` attributes can use the same function as argument. The arguments and return values for each function are different (in this case u16 and string and string and u16). The following should work:
```cpp
struct SpriteDimensionValues{
u16 X [[format("format_u16")]];
u16 Y [[format("format_u16")]];
};

fn format_u16(auto val) {
   return std::format("{:#x}",~val); //returns a string with ~val in hexadecimal
};
```
'format_write' function takes a string as input and returns a u16.

**Reply by Simran-B** [reply]
I'd appreciate it if you would test things before assuming they can't work.

https://github.com/WerWolv/ImHex/assets/7819991/27ce80d4-509d-4d9e-9332-45225e23840d

**Comment by paxcut** [comment]
I'll apologize but in my defense I never said or implied that the code didn't work. Yes, you can also return a pattern instead of a string, but in this case that creates an additional problem as noted so it seems that returning a string would be preferred. Not only  that would require  no casting, but it also gives you more control on the output (for example you could write "~value = {}" to indicate that the data displayed is the binary neg of the data at that variable, or maybe you want to display the value in binary, ...)
